src/main/java/bio/terra/service/filedata/azure/util/BlobContainerClientFactory.java

```diff
@@ -0,0 +1,187 @@
+
+    if (StringUtils.isBlank(accountKey)) {
+      throw new IllegalArgumentException("The account key is either null or empty.");
+    }
+
+    if (StringUtils.isBlank(containerName)) {
+      throw new IllegalArgumentException("The container name is either null or empty.");
+    }
+
+    blobContainerClient =
+        createBlobServiceClientUsingSharedKey(accountName, accountKey)
+            .getBlobContainerClient(containerName);
+    sasTokenGeneratorStrategy = SasTokenGeneratorStrategy.SharedKey;
+    blobContainerSasTokenCreds = null;
+  }
+
+  public BlobContainerClientFactory(
+      String accountName, TokenCredential azureCredential, String containerName) {
+
+    if (StringUtils.isBlank(accountName)) {
+      throw new IllegalArgumentException("The account name is either null or empty.");
+    }
+
+    if (StringUtils.isBlank(containerName)) {
+      throw new IllegalArgumentException("The container name is either null or empty.");
+    }
+
+    blobServiceClient =
+        createBlobServiceClientUsingTokenCredentials(
+            accountName,
+            Objects.requireNonNull(azureCredential, "Azure token credentials are null."));
+    blobContainerClient = blobServiceClient.getBlobContainerClient(containerName);
+
+    sasTokenGeneratorStrategy = SasTokenGeneratorStrategy.UserDelegatedKey;
+    blobContainerSasTokenCreds = null;
+  }
+
+  public BlobContainerClientFactory(String containerURLWithSasToken) {
+
+    BlobUrlParts blobUrl = BlobUrlParts.parse(containerURLWithSasToken);
+
+    blobContainerSasTokenCreds =
+        new AzureSasCredential(blobUrl.getCommonSasQueryParameters().encode());
+    blobContainerClient =
+        new BlobContainerClientBuilder()
+            .httpClient(httpClient)
+            .endpoint(
+                String.format(
+                    Locale.ROOT,
+                    "https://%s/%s",
+                    blobUrl.getHost(),
+                    blobUrl.getBlobContainerName()))
+            .credential(blobContainerSasTokenCreds)
+            .buildClient();
+
+    sasTokenGeneratorStrategy = SasTokenGeneratorStrategy.ContainerSasToken;
+  }
+
+  public BlobContainerClient getBlobContainerClient() {
+    return blobContainerClient;
+  }
+
+  public String createReadOnlySasUrlForBlob(String blobName) {
+    if (sasTokenGeneratorStrategy.equals(SasTokenGeneratorStrategy.SharedKey)) {
+      return createReadOnlySasUrlForBlobUsingClient(blobName);
+    }
+
+    if (sasTokenGeneratorStrategy.equals(SasTokenGeneratorStrategy.ContainerSasToken)) {
+      return createReadOnlySasUrlForBlobUsingContainerSas(blobName);
+    }
+
+    if (sasTokenGeneratorStrategy.equals(SasTokenGeneratorStrategy.UserDelegatedKey)) {
+      return createReadOnlySasUrlForBlobUsingUserDelegatedSas(blobName);
+    }
+
+    throw new RuntimeException("Failed to generate SAS token for blob. Invalid strategy set.");
+  }
+
+  private synchronized String createReadOnlySasUrlForBlobUsingUserDelegatedSas(String blobName) {
```

Comment:
Why is this `synchronized`? Do we expect this object to be used by multiple threads?

Reply:
BlobContainerClientFactory is thread-safe, which will allow creating instances of the factory as Beans that get wired up at start-up or as static references when needed.  Access to the cached key must be synchronized to manage its lifecyle. The key is cached for performance reasons (getting the key may require up to two calls, for each blob in the copy operation). But also, because getting the key is an auditable event that must be monitored for security reasons. Reducing the number of times, it occurs minimizes noise and makes it easier to detect unexpected behaviors/signals. 

I will refactor the management of the key into its own method to make it explicit.

Reply:
Is there a benefit to doing this on demand, instead of doing it during object construction? That would let us make the filed `final` and void the need for `synchronized`.

Reply:
I can't think of an easy way to do it in the constructor. The copier enumerates the list of source blobs and for each one a SAS URL is created on-demand at the beginning of the copy operation.

Reply:
I see, I missed how the key is automatically refreshed if it's expired. So it can't be `final` as it may change during this object's lifetime, and it can't always be generated because it's expensive to create. Putting it in a method that encapsulates this process and gives us a places to document it is the best we can do.

Reply:
This is fixed in the last update.